aim/db/migration/alembic_migrations/versions/dd2f91cf1b1e_set_nat_bd_ip_learn.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2017 Cisco Systems
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""Fix default value for BD's Limit IP Learning to subnets
+
+Revision ID: dd2f91cf1b1e
+Revises: 593d228d2fb4
+Create Date: 2017-11-06 16:39:33.713695
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = 'dd2f91cf1b1e'
+down_revision = '593d228d2fb4'
+branch_labels = None
+depends_on = None
+testing = True
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.sql.expression import update
+
+AIM_BRIDGE_DOMAINS = 'aim_bridge_domains'
+AIM_BRIDGE_DOMAINS_L3OUTS = 'aim_bridge_domain_l3outs'
+
+# A model of the BridgeDomainL3Out table with the fields we're interested
+bd_l3outs = sa.Table(AIM_BRIDGE_DOMAINS_L3OUTS, sa.MetaData(),
+                     sa.Column('bd_aim_id',
+                               sa.Integer,
+                               sa.ForeignKey('aim_bridge_domains.aim_id'),
+                               primary_key=True),
+                     sa.Column('name', sa.String(64)))
+
+# A model of the BridgeDomain table with the fields we're interested
+bridge_domains = sa.Table(AIM_BRIDGE_DOMAINS, sa.MetaData(),
+                          sa.Column('name', sa.String(64)),
+                          sa.Column('tenant_name', sa.String(64)),
+                          sa.Column('display_name', sa.String(64)),
+                          sa.Column('monitored', sa.Boolean,
+                                    nullable=False, default=False),
+                          sa.Column('aim_id',
+                                    sa.Integer,
+                                    primary_key=True, autoincrement=True),
+                          sa.Column('limit_ip_learn_to_subnets',
+                                    sa.Boolean))
+
+
+def upgrade():
+    session = sa.orm.Session(bind=op.get_bind())
+    my_bd_l3outs = get_bd_l3out_values(session)
+    for bd_l3out in my_bd_l3outs:
+        bd_update = update(bridge_domains).values(
```

Comment:
I wonder if this update could be done as a join between the BD table and BD-L3out table ... or possibly with a IN query - SQLAlchemy equivalent of
UPDATE aim_bridge_domain SET limit_ip_learn_to_subnets = 1 WHERE aim_id IN (SELECT DISTINCT bd_aim_id FROM aim_bridge_domain_l3outs)

Reply:
Just trying to make sure I understand what you're optimizing here -- are you thinking this can be done with a single update clause (i.e. without iterating over the bd_l3out entries)?

Reply:
Oh -- I think I see what you're saying. Rather than collect and iterate over the entries from the aim_bridge_domain_l3outs table, just create the aim_bridge_domains row updates from a single query via the join?

Reply:
I tried creating the above, but I run into problems with the Foreign Key. I tried fixing this by adding the relationship to the BridgeDomain table definition, but had trouble getting it to work. If this patch wasn't just for DB migrations (i.e. normal operational code path), I'd be inclined to spend more time figuring out how to make this work, but given that it's just for DB upgrades (and works), I'd prefer to leave this as a potential future optimization for now (maybe add a REVISIT?).